### PR TITLE
WIP: Optimize `ComputeLastIndex`

### PR DESCRIPTION
**Pull request by xtqqczze**
https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs,213

### CODE REVIEW

**Comment by danmoseley** [comment]
Cc @MihaZupan

**Comment by MihaZupan** [comment]
Thanks Dan, I wasn't aware of this PR.
@xtqqczze did you see any perf improvements from replacing the substraction with an xor like this?

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs [hunk 1]
```diff
@@ -934,6 +934,15 @@ private static uint Crc32CCore(ref uint lookupTable, uint crc, uint data)
             }
         }
 
+        internal static int Log2NonZero(uint value)
+        {
+            Debug.Assert(value != 0);
+
+            // Leading zero count is in the range 0..31, so the identity "31 ^ n == 31 - n" holds.
+            // Subtraction from a constant requires one more instruction.
+            return 31 ^ LeadingZeroCount(value);
```
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:943]: Ideally the JIT could recognize these pattern and just emit `bsr` (on x86) like [gcc and clang do](https://godbolt.org/z/xhPq3E8K3).

For reference: [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgBEBLbAcwDsJcGDmFwBuGvSasAcgFd8MKMLE0JAZiakGAYQYBvGgyMMOvDIt7YANlKQmzDADIRupaRF4AtRRAAUs0wwGADdrWRgASkNjA2pjeIY2GGBZbhYAQVxcRQxfUKtwhgBCAF4GOgjxamiEhgB6OqcYbAATU24GAC8fBkhZBw5cewYMAAsYBihsXm4JuhYWNXI0BlwIEfGTFpgzDgwATwYAIiWGAD0GXgYSstO4S6OGUYgrFtwWGoSGhgBlWWAMFMwEIPAwAGZQCD4BjYXoeQTTIKwACOAVgQw8E3w0AmpkEUFkwI4Hg+cVqTAA7AxThcAEJ7ADyAAdFNgQbx3o5mm0Zt5IdoIP1cvlwpVPgBfT6fQIWay2exBZyudxeHzkfyBEJhSKfWLkxLJVIZLI5PLa4plCpVT7xb5c1rtLo9PoDIamDYTKYzOYLJYrNYerY7IQHY40y7XW7kBj3XiPZ6vd424zfP4AoHs8GQ6GwyAcjCIyYwVEcdEMTEMbGwez4wns0n64hUu4MekYZms9mc7ntPkQAVCs0FSJVeKS6jioA)

Comment (xtqqczze) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:943]: Indeed [gcc and clang](https://godbolt.org/z/MzTedYYo3) emit `lzcnt` and `zor` when specifying `march=znver4`.

**Pull request closed without merging**